core/src/main/scala/spinal/core/sim/SpinalSpec.scala

```diff
@@ -0,0 +1,273 @@
+    def buildBench: SimCompiled[Dut] = config.compile(dut)
+
+    /** Has this testbench already been compiled for caching? */
+    private var triedCompile: Boolean = false
+
+    /** The cached bench. Can be `None` if `!triedCompile` or if compile failed */
+    private var cachedBench: Option[SimCompiled[Dut]] = None
+
+    /** Set of functions to build text for test results */
+    private object txt {
+      def itCompiles: String = s"$getName compiles"
+      def notCompiles: String = s"$getName does not compile"
+      def itShould(doWhat: String): String = s"$getName should $doWhat"
+      def test(doWhat: String): String = s"$getName test: $doWhat"
+    }
+
+    /** Compile `cachedBench` if caching is enabled and not compiled yet
+      *
+      * Use this function only before tests, NEVER in a test
+      */
+    protected def prepareCachedBench(): Unit = {
+      if (caching && !triedCompile) {
+        triedCompile = true
+        txt.itCompiles in {
+          cachedBench = Some(buildBench)
+        }
+      }
+    }
+
+    /** Function used to run a test/simulation
+      *
+      * @param name The name of the test/simulation
+      * @param f doSim function, eg: `_.doSimUntilVoid(name)` or `dut => dut.doSim(name, seed)`
+      * @param body the code to run on the (possibly abstracted) `dut` during the simulation
+      */
+    protected def sim(name: String, f: SimCompiled[Dut] => (Dut => Unit) => Unit)(
+        body: Body
+    ): Unit = {
+      prepareCachedBench()
+      name in {
+        val bench =
+          if (caching) cachedBench getOrElse cancel
+          else buildBench
+        f(bench) { dut =>
+          val softDut = preSimTransform(dut)
+          body(softDut)
+        }
+      }
+    }
+
+    /** Helper to call sim with optional seed */
+    final protected def sim(name: String, seed: Option[Int]): Body => Unit = seed match {
+      case Some(s) => sim(name, _.doSim(name, s))
+      case None    => sim(name, _.doSim(name))
+    }
+
+    /** Helper to call sim with doSimUntilVoid and optional seed */
+    final protected def simUntilVoid(name: String, seed: Option[Int]): Body => Unit = seed match {
+      case Some(s) => sim(name, _.doSimUntilVoid(name, s))
+      case None    => sim(name, _.doSimUntilVoid(name))
+    }
+
+    /** Context for `it should "do something"` */
+    class Should(doWhat: String, seed: Option[Int] = None) {
+
+      /** Specify a seed for this test */
+      def withSeed(seed: Int): Should = new Should(doWhat, Some(seed))
+
+      /** Specify a seed for this test, or to not fix one */
+      def withSeed(seed: Option[Int]): Should = new Should(doWhat, seed)
+
+      /** Run a simulation to check that it does the thing it should
+        *
+        * {{{
+        * it should "do something" in { dut =>
+        *   // code to check that it does the thing it should
+        * }
+        * }}}
+        */
+      def in: Body => Unit = sim(txt.itShould(doWhat), seed)
```

Comment:
I agree but I couldn't find a better one too…

(Hmm, what about `inSim`/`inSimUntilVoid`?)

Reply:
why not `sim` and `simUntilVoid` (if we want to stick with the `UntilVoid` postfix)?

Reply:
Hmm, shorter than `inSim` and `isSimUntilVoid` but the sentence is less `Spec`-like: "it should work sim" instead of "it should work in sim"… I don't know, what do you prefer?

Reply:
also sounds ok to me